### PR TITLE
IProviderAdapter and DefaultProviderAdapter

### PR DESCRIPTION
**Pull request by andxu**
Signed-off-by: andxu <andxu@microsoft.com>

### CODE REVIEW

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package org.eclipse.jdt.ls.debug.adapter;
+
+public interface IProviderAdapter {
```
Comment (testforstephen) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java]: IProviderManager

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+public class DefaultProviderAdapter implements IProviderAdapter {
+
+    private Map<Class<? extends IProvider>, IProvider> providerMap;
+    
+    public DefaultProviderAdapter() {
+        providerMap = new HashMap<>();  
+    }
+    
+    @SuppressWarnings("unchecked")
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java]: Please provide a reason for suppressing the specified warning.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package org.eclipse.jdt.ls.debug.adapter.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.jdt.ls.debug.adapter.IProvider;
+import org.eclipse.jdt.ls.debug.adapter.IProviderAdapter;
+
+public class DefaultProviderAdapter implements IProviderAdapter {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java]: If it's default, then just use the name "ProviderAdapter".

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java (hunk 1 above)
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java]: Please use this to replace IProviderContext but keep the name IProviderContext.

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IConfigurable.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * Contributors:
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.jdt.ls.debug.adapter;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public interface IConfigurable {
```
Comment (akaroml) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IConfigurable.java]: Please merge this with IProvider and keep the name IProvider.

**Review by akaroml — CHANGES_REQUESTED** [review]

**Review by akaroml — DISMISSED** [review]

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IConfigurable.java (hunk 4 above)
Comment (andxu) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IConfigurable.java]: done

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java (hunk 1 above)
Comment (andxu) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderAdapter.java]: Done

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java (hunk 2 above)
Comment (andxu) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java]: inline  comments added

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java (hunk 3 above)
Comment (andxu) [reply org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/impl/DefaultProviderAdapter.java]: Done

File: org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderContext.java [hunk 5]
```diff
@@ -1,19 +1,23 @@
...
+ *     Microsoft Corporation - initial API and implementation
+ *******************************************************************************/
 
 package org.eclipse.jdt.ls.debug.adapter;
 
 public interface IProviderContext {
+    
+    <T extends IProvider> T getProvider(Class<T> clazz);
+
+    void registerProvider(Class<? extends IProvider> clazz, IProvider provider);
```
Comment (ansyral) [inline org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/IProviderContext.java:18]: nice!

**Review by ansyral — APPROVED** [review]
:shipit:

**Review by akaroml — APPROVED** [review]

**Pull request merged**